Don't force a new resource when updating early renewal time.

anuraaga opened the pull request:
Early renewal time is not a feature of a certificate and just used by terraform to determine whether to recreate a cert. Updating the time itself doesn't need to update the cert. If users used the wrong value for early renewal (for example, specifying a time from issuance instead of the correct time from expiration), they will be forced to rotate certs instead of just tweaking this value.

Also updates `go.mod` / `vendor` - couldn't get it to build when it depended on the alpha SDK.

anuraaga commented:
@mildwonkey Hope you can take a look at this PR, otherwise we'll have to directly much with our state since we messed up the value :-)

anuraaga commented:
@mildwonkey Ping - would be cool if you get a chance to look at this. Thanks!

apparentlymart commented:
Hi @anuraaga! Thanks for working on this and sorry for the slow response.

The Terraform team is currently 100% focused on finishing up the v0.12.0 release so we probably won't be able to look at this for a few more weeks at least. Unfortunately I think directly editing the state here is probably the most expedient path, so that our work on the v0.12.0 release isn't a blocker for you to be able to move forward with `terraform apply`. :confounded:

anuraaga commented:
Actually we still have about half a year (instead of the intended 8 and a half years) before the certs get renewed so we may be able to wait until after 0.12.0. Thanks for the note :)

anuraaga commented:
@apparentlymart Congratulations on the release of 0.12 :) Just wanted to bring this back in case it's possible to take a look now.

apparentlymart commented on tls/resource_certificate.go:
```diff
@@ -186,25 +185,45 @@ func DeleteCertificate(d *schema.ResourceData, meta interface{}) error {
+	if certNeedsRenewal(d) {
+		// Treat an expired certificate as not existing, so we'll generate
+		// a new one with the next plan.
+		d.SetId("")
```
I think the effect of this would be to return to Terraform Core with no certificate at all after update, which Terraform Core v0.12 will then consider an error (an update plan cannot implicitly become a destroy during apply).

I think there's a different way to get the result we want here, though:

The additional `*Resource` field `CustomizeDiff` was added in the meantime since this function was originally implemented, and it allows (among other things) the use of logic to conditionally set `ForceNew` on an attribute. I think it should be possible to make the `CustomizeDiff` function be responsible for handling `early_renewal_hours` and `validity_end_time`, by checking them and then calling [`d.ForceNew`](https://godoc.org/github.com/hashicorp/terraform/helper/schema#ResourceDiff.ForceNew) with whichever of the two attributes is the one detected as triggering renewal.

That approach would avoid the need to handle this conditionally in `Read` (it can just be a no-op function) and would also have a nice usability benefit: the rendered Terraform plan will show an expired resource as a "must be replaced" plan, rather than the old certificate just silently vanishing during refresh and being planned for creation.

anuraaga commented:
@apparentlymart Thanks for the advice - I've updated to use `CustomizeDiff` and added test cases that seem to verify the behavior. Haven't written such large change in the terraform codebase before though welcome to any style, etc suggestions.

Also, while I have your attention, if I can point you at this completely unrelated PR for possibly a push... 

https://github.com/terraform-providers/terraform-provider-kubernetes/pull/389

It's been approved for a while and is quite critical for using terraform with k8s - I ended up forking the provider since I couldn't upgrade to 0.12 without downtime since parser whitespace changes were triggering a destroy instead of update ><

apparentlymart commented:
Thanks for working on these updates @anuraaga! This looks great to me. I'm going to merge it now.

---

(On the subject of the Kubernetes provider PR: I'm not on the team for that provider and so I can't influence that team's prioritization. It looks like one of the team members is already participating and has been repeatedly tagged on it, so I'm sure it's on the team's radar and they'll get to it.)
